### DISCUSSION TITLE
expand_selection should take comments into account (like VS Code does)

### DISCUSSION THREAD

**Discussion by jasonwilliams** [helix-editor/helix#8888] [Ideas]
I'm going to assume this may be some tree sitter question rather than the internals of Helix itself.

When using `expand_selection` its useful to include the comment as part of the same block, this is because you may select the function to move it somewhere else, but the problem right now is the comment doesn't move with it. 

This is something I think VS Code gets right because the expansion treats the comment above as part of the function.

VS Code selects the comment above the function as part of the expansion
![vscodeWorking](https://github.com/helix-editor/helix/assets/936006/10f98165-d8d7-402a-b69e-42765533147e)

Helix does not recognise the comment
![helixNotWorking](https://github.com/helix-editor/helix/assets/936006/aa611cfa-ae04-4513-99c5-b19a805b6182)

Can this be fixed in Helix or does it need to be applied somewhere else?

**Comment by the-mikedavis** [comment]
`expand_selection` finds which node in the syntax tree is covered by the current selection and sets the selection to cover the parent node. The comment isn't included here because it doesn't belong to the same subtree as the function. You could hypothetically change the typescript tree-sitter grammar to 'attach' comments to the next function definition but I believe tree-sitter-typescript's current behavior is correct - comments don't really syntactically belong to the functions.

**Reply by jasonwilliams** [reply]
Thanks for the response @the-mikedavis 

I understand a function will be one node in the AST whilst a comment may be another, and out of the box (without doing any work), the expansion wouldn’t choose the comment because it’s seen as a sibling. 

That being said, I do believe VS Code have got this right and I’ll explain why.

- If you delete a function, the block comment directly above it no longer makes sense.
- If you move a function, you’ll need to move the comment with it anyway
- If you use JSdoc, that will be in the comment above (which typescript uses)
